### PR TITLE
[49] Groups CRUD

### LINKED ISSUE #49: Students Can Form Groups
~~Blocked on #36.~~

### PR DESCRIPTION
**Pull request by esoterik**
Resolves #49. 

Still needs a bit of documentation but is otherwise ready for review!

### CODE REVIEW

File: app/models/group.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+#
+# Model for Housing Groups
+#
+# @attr [Integer] size The room size that the Group wants.
+# @attr [String] status The state of the group (open, full, or locked)
+# @attr [User] leader The student that represents the Group.
+# @attr [Array<User>] members The members of the group, excluding the leader.
+# @attr [Draw] draw The Draw that this Group is in.
+class Group < ApplicationRecord
```
Comment (esoterik) [inline app/models/group.rb]: this is missing the more complex validations (no more students than size, all students must be in the same draw, size must be available in the current draw, etc), but I think those should be done in a separate issue because this PR is already fairly long

File: spec/policies/group_policy_spec.rb [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        it { is_expected.not_to permit(user, Group) }
+      end
+      context 'not in draw' do
+        it { is_expected.not_to permit(user, Group) }
+      end
+    end
+    permissions :index? do
+      it { is_expected.to permit(user, [group, other_group]) }
+    end
+    permissions :show? do
```
Comment (orenyk) [inline spec/policies/group_policy_spec.rb]: Are we allowing users to view groups in other draws?

File: app/services/creators/group_creator.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  def add_draw_to_params
+    @params.merge!(draw_id: User.find(params[:leader_id]).draw.id)
+  end
+
+  def remove_blank_members
+    @params[:member_ids] = params[:member_ids].reject(&:empty?)
+  end
+
+  def success(group)
+    { object: [group.draw, group], group: group,
```
Comment (orenyk) [inline app/services/creators/group_creator.rb]: I was confused about this value for `object` but now I see you're using it to redirect to a nested resource, nice!

File: spec/services/creators/group_creator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    end
+  end
+
+  it 'does not create when given invalid params' do
+    params = instance_spy('ActionController::Parameters', to_h: {})
+    expect(described_class.new(params).create![:object]).to be_nil
+  end
+
+  def params_hash
+    draw = FactoryGirl.create(:draw_with_members, students_count: 1)
```
Comment (orenyk) [inline spec/services/creators/group_creator_spec.rb]: Do we want to memoize these just in case this ever gets called twice? I know it's unlikely but seeing guaranteed resource creation within a helper method is a little troubling to me.

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -23,7 +23,7 @@ def index
   # @param [String] action The action to render when no object passed.
   #   (Creation / update failure, destruction success)
   # @param [String] path The path to redirect to when no object passed.
-  def handle_action(object:, msg:, action: nil, path: nil)
+  def handle_action(object:, msg:, action: nil, path: nil, **_)
```
Comment (orenyk) [inline app/controllers/application_controller.rb]: Why was this necessary?

Comment (esoterik) [inline app/controllers/application_controller.rb]: The method gets passed the splatted `results` from creator objects; this allows those results to have arbitrary keys that get used in the actual controller method (here, `:group` is the key that i used)

File: app/controllers/groups_controller.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  def set_group
+    @group = Group.find(params[:id])
+  end
+
+  def set_draw
+    @draw = Draw.find(params[:draw_id])
+  end
+
+  def set_user
+    @user = current_user
```
Comment (esoterik) [inline app/controllers/groups_controller.rb]: this is actually totally unnecessary because `current_user` is available in views, will take it out

File: app/models/user.rb [hunk 7]
```diff
@@ -7,6 +7,7 @@ class User < ApplicationRecord
          :recoverable, :rememberable, :trackable, :validatable
 
   belongs_to :draw
+  belongs_to :group
```
Comment (esoterik) [inline app/models/user.rb]: hmm, I'm not sure. 

At the moment, leaders are not included in the members of a draw, so there's no conflict. Do you think it will help us in the long run to duplicate leaders in the members array? I'm still not sure that we would need the `has_one` because it would always be the same id as the `belongs_to`.

File: app/services/updaters/group_updater.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+# Service object to update groups
```
Comment (esoterik) [inline app/services/updaters/group_updater.rb]: I think we're somewhat inconsistent about this across the board; I think that the empty line is probably better so I'll add it

File: app/views/groups/_form.html.erb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+<%= simple_form_for [@draw, @group] do |f| %>
```
Comment (esoterik) [inline app/views/groups/_form.html.erb]: not 100% sure, simple_form figures it out I think. (see this [stackoverflow](http://stackoverflow.com/questions/23945600/rails-4-action-routes-with-simple-form-and-shallow-nested-resources))

File: app/policies/group_policy.rb [hunk 10]
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+# Class for Group permissions
+class GroupPolicy < ApplicationPolicy
+  def show?
+    true
+  end
+
+  def index?
```
Comment (esoterik) [inline app/policies/group_policy.rb]: yeah, I think that the index actions will end up being more complicated so they belong in separate issues.

File: db/migrate/20161224170703_create_groups.rb [hunk 11]
```diff
@@ -0,0 +1,15 @@
+class CreateGroups < ActiveRecord::Migration[5.0]
+  def change
+    create_table :groups do |t|
+      t.integer :size, null: false, default: 1
+      t.integer :status, null: false, default: 0
+      t.belongs_to :leader, index: true, null: false
```
Comment (esoterik) [inline db/migrate/20161224170703_create_groups.rb]: Yeah, I was thinking that the leader would always be a student because they eventually get the permission to select a suite for the group.

File: spec/policies/group_policy_spec.rb (hunk 2 above)
Comment (esoterik) [inline spec/policies/group_policy_spec.rb]: I think it's useful for trying to predict what room sizes will be leftover for your draw

**Comment by esoterik** [comment]
pushed an update!

File: app/models/user.rb (hunk 7 above)
Comment (orenyk) [inline app/models/user.rb]: I don't know if Rails will know how to assign the `group_id` for the `leader` unless they're included in the `members` array, since this `belongs_to` corresponds to the `has_many :members` in the Group model, whereas a `has_one :group` will correspond to the `belongs_to :leader` in the Group model. Now, if we include the leader as a member (probably through a callback or something), then that will provide the `leader.group` method and we shouldn't need to use `has_one :group` here, and if we did include it anyway I'm not sure if it would break anything.

Does that make sense?

File: app/views/groups/_form.html.erb (hunk 9 above)
Comment (orenyk) [inline app/views/groups/_form.html.erb]: Ohhh I didn't realize that it was a nested route when I first looked here, so that's why I was confused. Now it makes sense.

File: db/migrate/20161224170703_create_groups.rb (hunk 11 above)
Comment (orenyk) [inline db/migrate/20161224170703_create_groups.rb]: Well, I'm guessing that in some situations the admins would be performing suite selection as well, but yea we're going to want them to assign a leader. We just have to make sure that it's easy to do.

File: spec/policies/group_policy_spec.rb (hunk 2 above)
Comment (orenyk) [inline spec/policies/group_policy_spec.rb]: Yea, but do they need to see which students are in each group or just the number of groups?

Comment (orenyk) [inline spec/policies/group_policy_spec.rb]: I mean, if it's not sensitive info (who's living with who) then it doesn't matter, and we can always change it later anyway.

File: app/controllers/application_controller.rb (hunk 5 above)
Comment (orenyk) [inline app/controllers/application_controller.rb]: Ohhh otherwise it was throwing some kind of mismatched parameter error? Got it.

File: app/models/group.rb [hunk 12]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+#
+# Model for Housing Groups
+#
+# @attr [Integer] size The room size that the Group wants.
```
Comment (orenyk) [inline app/models/group.rb]: So I think you've been doing YARD wrong 😕 - I'm pretty sure the format is `@attr attr_name [Type] description`, and until now the attribute name has been going _after_ the type. Let's fix all the new comments in this branch and I'll open an issue to update all our earlier documentation at some point.

File: spec/factories/groups.rb [hunk 13]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+FactoryGirl.define do
+  factory :group do
+    size 1
+    status 'open'
```
Comment (orenyk) [inline spec/factories/groups.rb]: Shouldn't this actually be closed by default because a Group of size 1 is just the leader?

File: app/models/user.rb (hunk 7 above)
Comment (esoterik) [inline app/models/user.rb]: yup, will do!

File: spec/policies/group_policy_spec.rb (hunk 2 above)
Comment (esoterik) [inline spec/policies/group_policy_spec.rb]: yeah, I think we can change this later if necessary

File: app/models/group.rb (hunk 12 above)
Comment (esoterik) [inline app/models/group.rb]: ohno! I must have looked at an inaccurate reference when I started using it and never checked up on it

**Comment by esoterik** [comment]
ok, should be good to go now!

File: app/controllers/groups_controller.rb [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+  def show; end
+
+  def new
+    @group = Group.new(draw: @draw)
+  end
+
+  def create
+    p = group_params.to_h
+    p[:leader_id] = current_user.id unless current_user.admin?
+    result = GroupCreator.new(p).create!
```
Comment (orenyk) [inline app/controllers/groups_controller.rb]: I know this is probably a minor point, but aren't we passing in a `Hash` here instead of an `ActionController::Parameters` object as we've done in the other controllers / creators (and as we document in the `GroupCreator`)? With duck typing it shouldn't matter, but we should update the documentation in the `GroupCreator` to be accurate.

File: app/services/creators/group_creator.rb [hunk 15]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+#
+# Service object to create Groups.
+class GroupCreator < Creator
+  # Initialize a new GroupCreator
+  #
+  # @param params [ActionController::Parameters] The params object from
```
Comment (orenyk) [inline app/services/creators/group_creator.rb]: See my last comment.

File: spec/factories/groups.rb [hunk 16]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+FactoryGirl.define do
+  factory :group do
+    size 1
+    status 'full'
+    association :leader, factory: :student_in_draw
+    draw { leader.draw }
```
Comment (orenyk) [inline spec/factories/groups.rb]: Do we need to add `members { [leader] }` as well, or do you want to do that when we add the validations / callbacks that ensure that the leader is also a member?

File: app/controllers/groups_controller.rb (hunk 14 above)
Comment (esoterik) [inline app/controllers/groups_controller.rb]: oooh, turns out YARD has a way to [denote](http://yardoc.org/types.html#see_results) duck types so I'll update it accordingly

File: spec/factories/groups.rb (hunk 16 above)
Comment (esoterik) [inline spec/factories/groups.rb]: I added the callback already but not the validation, so I think we're okay

File: app/controllers/groups_controller.rb (hunk 14 above)
Comment (orenyk) [inline app/controllers/groups_controller.rb]: Excellent, nice find 😄

File: spec/factories/groups.rb (hunk 16 above)
Comment (orenyk) [inline spec/factories/groups.rb]: Whoops I missed that, I'll go double check and then we should be good to go.

File: app/models/group.rb [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+  after_create :add_leader_to_members
+
+  def name
+    "#{leader.name}'s Group"
+  end
+
+  private
+
+  def add_leader_to_members
+    members.append(leader)
```
Comment (orenyk) [inline app/models/group.rb]: Hmm, do we want to maybe make this an `after_save` and ensure that the leader is _always_ a member? Also, if the leader was somehow already selected as a member (e.g. admin-created group) would this result in a doubled reference?

**Comment by esoterik** [comment]
callback is now an `after_save` and checks if the leader is already a member before appending!
